src/Raven.Server/Smuggler/Documents/DatabaseSource.cs

```diff
@@ -151,19 +178,48 @@ public IEnumerable<DocumentItem> GetRevisionDocuments(List<string> collectionsTo
             if (revisionsStorage.Configuration == null)
                 yield break;
 
-            var documents = collectionsToExport.Count != 0
-                ? revisionsStorage.GetRevisionsFrom(_context, collectionsToExport, _startDocumentEtag, int.MaxValue)
-                : revisionsStorage.GetRevisionsFrom(_context, _startDocumentEtag, int.MaxValue);
+            var enumerator = new PulsedTransactionEnumerator<Document, DocumentsIterationState>(_context,
+                state =>
+                {
+                    if(collectionsToExport.Count != 0)
+                        return GetRevisionsFromCollections(_context, collectionsToExport, state);
+
+                    return revisionsStorage.GetRevisionsFrom(_context, state.StartEtag, int.MaxValue);
+                },
+                new DocumentsIterationState(_context) // initial state
+                {
+                    StartEtag = _startDocumentEtag,
+                    StartEtagByCollection = collectionsToExport.ToDictionary(x => x, x => _startDocumentEtag)
+                });
 
-            foreach (var document in documents)
+            while (enumerator.MoveNext())
             {
                 yield return new DocumentItem
                 {
-                    Document = document
+                    Document = enumerator.Current
                 };
             }
         }
 
+        private IEnumerable<Document> GetRevisionsFromCollections(DocumentsOperationContext context, List<string> collections, DocumentsIterationState state)
+        {
+            foreach (var collection in collections)
```

Comment:
Let's use the `state.StartEtagByCollection` for this, instead of keeping the data in two locations